src/github_client.rs

```diff
@@ -30,39 +31,49 @@ impl GithubClient {
     }
 
     /// Gets all notifications since a timestamp even if they've been read
-    fn get_notifications(&self, since: DateTime<Local>) -> Result<Response, Error> {
+    async fn get_notifications(&self, since: DateTime<Local>) -> Result<Response, Error> {
         self.client
             .get(GITHUB_API_URL)
             .basic_auth(&self.username, Some(&self.token))
             .query(&[("since", since.to_rfc3339()), ("all", String::from("true"))])
             .send()
+            .await
     }
 
     /// Gets a response from an API URL
-    fn get(&self, api_url: &str) -> Result<Response, Error> {
+    async fn get(&self, api_url: &str) -> Result<Response, Error> {
         self.client
             .get(api_url)
             .basic_auth(&self.username, Some(&self.token))
             .send()
+            .await
     }
 
     /// Gets an HTML URL from a notification's API URL and bundles them together
-    fn get_html_url(&self, notification: GithubNotification) -> Result<NotificationWithUrl, Error> {
-        let response = self.get(&notification.subject.url)?;
-        let url: HasHtmlUrl = response.json::<HasHtmlUrl>()?;
+    async fn get_html_url(
+        &self,
+        notification: GithubNotification,
+    ) -> Result<NotificationWithUrl, Error> {
+        let response = self.get(&notification.subject.url).await?;
+        let url: HasHtmlUrl = response.json::<HasHtmlUrl>().await?;
         Ok(NotificationWithUrl::new(notification, url))
     }
 
     /// Gets all notifications since a timestamp and bundles them with a human-usable HTML URL
-    pub fn fetch_notifications(
+    pub async fn fetch_notifications(
         &self,
         since: DateTime<Local>,
     ) -> Result<Vec<NotificationWithUrl>, Error> {
-        let response = self.get_notifications(since)?;
-        let notifications: Vec<GithubNotification> = response.json::<Vec<GithubNotification>>()?;
-        notifications
-            .into_iter()
-            .map(|notification| self.get_html_url(notification))
-            .collect()
+        let response = self.get_notifications(since).await?;
+        let notifications: Vec<GithubNotification> =
+            response.json::<Vec<GithubNotification>>().await?;
+        join_all(
+            notifications
+                .into_iter()
+                .map(|notification| self.get_html_url(notification)),
+        )
+        .await
+        .into_iter()
+        .collect()
```

Comment:
There's probably a much better way of doing this—preferably one without `join_all`—but I'll have to spend time actually thinking about what's going on here instead of hammering at the keyboard until it builds to find out.

Reply:
test comment